src/main/java/net/logstash/logback/util/ReusableByteBuffer.java

```diff
@@ -0,0 +1,240 @@
+ * Buffers are disposed except the initial buffer which is reused by subsequent usage.
+ *
+ * @author brenuart
+ *
+ */
+public class ReusableByteBuffer extends OutputStream {
+
+    /**
+     * The default size of the initial buffer
+     */
+    public static final int INITIAL_SIZE = 1024;
+
+    /**
+     * Constant with an empty byte array
+     */
+    private static final byte[] EMPTY_BYTES = new byte[0];
+
+    /**
+     * The buffers used to store the content bytes
+     */
+    private final List<byte[]> buffers = new ArrayList<>();
+
+    /**
+     * The number of bytes already written in previous buffers (other than tail).
+     */
+    private int alreadyBufferedSize = 0;
+
+    /**
+     * The write index in the tail buffer
+     */
+    private int index = 0;
+
+    /**
+     * Is the stream closed?
+     */
+    private boolean closed = false;
+
+
+    /**
+     * Create a new {@link ReusableByteBuffer}
+     * with the default initial capacity of 1024 bytes.
+     */
+    public ReusableByteBuffer() {
+        this(INITIAL_SIZE);
+    }
+
+    /**
+     * Create a new {@link ReusableByteBuffer}
+     * with the specified initial capacity.
+     *
+     * @param initialBlockSize the initial buffer size in bytes
+     */
+    public ReusableByteBuffer(int initialBlockSize) {
+        if (initialBlockSize <= 0) {
+            throw new IllegalArgumentException("Initial block size must be greater than 0");
+        }
+        this.buffers.add(new byte[initialBlockSize]);
+    }
+
+
+    @Override
+    public void write(int datum) throws IOException {
+        if (this.closed) {
+            throw new IOException("Stream closed");
+        }
+
+        growIfNeeded();
+        getTailBuffer()[this.index++] = (byte) datum;
+    }
+
+
+    @Override
+    public void write(byte[] data, int offset, int length) throws IOException {
+        if (data == null) {
+            throw new NullPointerException();
+        }
+        if (offset < 0 || offset + length > data.length || length < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (this.closed) {
+            throw new IOException("Stream closed");
+        }
+
+
+        while (length > 0) {
+            byte[] buffer = getTailBuffer();
+            int freeSpace = buffer.length - this.index;
+
+            if (freeSpace > 0) {
+                int toCopy = Math.min(freeSpace, length);
+                System.arraycopy(data, offset, buffer, this.index, toCopy);
+                offset += toCopy;
+                this.index += toCopy;
+                length -= toCopy;
+            }
+
+            if (length > 0) {
+                growIfNeeded();
+            }
+        }
+    }
+
+
+    @Override
+    public void close() {
+        this.closed = true;
+    }
+
+
+    /**
+     * Return the number of bytes stored in this {@link ReusableByteBuffer}.
+     */
+    public int size() {
```

Comment:
The term "size" is currently used for two things... the number of bytes written to the buffer (here), and the initial size (in the constructor param and a field).

To differentiate the two, I'd suggest using the term "capacity" when referring to the allocated capacity of the buffer. And "size" when referring to the bytes actually written.

Reply:
Correct me if I'm wrong... what you propose is:
- keep the `size()` method as it is atm
- rename the `initialSize` constructor arg to `initialCapacity`

Is that correct?

Reply:
Yes.  I wrote this comment first in my review (since it's where I noticed size was being used for two things), before I mentioned size->capacity rename elsewhere.   Just wanted to explain my reasoning.

Reply:
Hmmm...
`CompositeJsonLayout` already had a `setMinBufferSize` method that is now reused to set the _initial capacity_ of the buffers. To make things clear, the method should have the same name in `CompositeJsonEncoder`...

Knowing that, do you still want me to change the `initialSize` of the ReusableBuffer to `initialCapacity`? It is an internal implementation after all - and it doesn't have to follow the same naming as the "public" interfaces represented by the two composite layout/encoder...

Reply:
Sigh.  That's unfortunate. 

But yeah, let's use capacity when referring to allocated capacity in the new code.  I think it's much more clear when dealing with allocated vs used numbers.